### PR TITLE
Append and created a testthat unit test of the fix

### PR DESCRIPTION
**Pull request by Peacesandy**
1. Appended * 1e6 to line 28 and made some modifications to the code base
2. create a testthat unit test of the fix on the file testthat.R

## Console output after running the code

Warning (test-lefser.R:14:3): lefser and lefserPlot work
Convert counts to relative abundances with 'relativeAb()'
Backtrace:
    ▆
 1. ├─testthat::expect_error(lefser(zellersub, groupCol = NULL, blockCol = NULL)) at test-lefser.R:14:3
 2. │ └─testthat:::quasi_capture(...)
 3. │   ├─testthat (local) .capture(...)
 4. │   │ └─base::withCallingHandlers(...)
 5. │   └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
 6. └─lefser::lefser(zellersub, groupCol = NULL, blockCol = NULL)

Warning (test-lefser.R:15:3): lefser and lefserPlot work
Convert counts to relative abundances with 'relativeAb()'
Backtrace:
    ▆
 1. ├─withr::with_seed(...) at test-lefser.R:15:3
 2. │ └─withr::with_preserve_seed(...)
 3. └─lefser::lefser(zellersub, groupCol = "study_condition", blockCol = NULL)

Warning (test-lefser.R:26:3): lefser and lefserPlot work
Convert counts to relative abundances with 'relativeAb()'
Backtrace:
    ▆
 1. ├─withr::with_seed(...) at test-lefser.R:26:3
 2. │ └─withr::with_preserve_seed(...)
 3. └─lefser::lefser(zellersub, groupCol = "study_condition", blockCol = "age_category")
───────────────────────────────────────────────────────────────────────────────────

══ Results ════════════════════════════════════════════════════════════════════════
Duration: 5.9 s

[ FAIL 0 | WARN 3 | SKIP 0 | PASS 15 ]

1.For the append task 

- The counts are divided by column sums and then multiplied by 1e6 to obtain relative abundances.

I'm open to corrections and feedbacks and possibility to restructure the code if it's not correct

### CODE REVIEW

**Review by lwaldron — CHANGES_REQUESTED** [review]
Great start @Peacesandy ! A few things:
- [ ] don't commit .Rdata or .Rhistory
- [ ] limit changes to relativeAb to those addressing the bug. The additional changes make it difficult to distinguish between changes in response the bug, changes addressing something else, and purely formatting changes. In general, all these should be kept in separate PRs. Note that in Bioconductor style we don't usually add blank lines. There also shouldn't usually be `print()` or `cat()` statements, we use `message()` to provide messaging while respecting when the user has settings to suppress messages.
- [ ] it looks like you might be multiplying by 1e6 twice? Make sure the output is correct, with values between 0 and 1e6. 
- [ ] Document that relativeAb returns counts per million 
- [ ] Change warning to check that colSums are 1e6 on lefser.R#L273: https://github.com/waldronlab/lefser/blob/db8f36ad48b05d5749518650ae490234cdfa7e1d/R/lefser.R#L273

to:
```r
    if (checkAbundances && !all.equal(colSums(relab_data), rep(1e6, ncol(relab_data)), check.attributes = FALSE))
```
- [ ] add unit tests. Here are some:
```r
expect_warning(lefser(zeller14sub, groupCol = "study_condition"), regexp = "^Convert counts.+")
expect_equal(assayNames(zeller14ra), c("rel_abs", "exprs"))
expect_equal(colSums(assay(zeller14ra, "rel_abs")), rep(1e6, ncol(zeller14ra)), check.attributes = FALSE)
set.seed(1)
expect_no_warning(x <- lefser(zeller14ra, groupCol = "study_condition"))
expect_equal(dim(x), c(110, 2))
expect_equal(names(x), c("Names", "scores"))
expect_equal(x[1, "Names"], "k__Bacteria|p__Firmicutes")
```

**Comment by lwaldron** [comment]
(note I just updated my comments)

**Comment by Peacesandy** [comment]
Okay, I'll make the changes and get back

**Comment by Peacesandy** [comment]
- I used the .gitignore file, to ignore the R.history and R.data
- I refactrored the codebase and made the changes
- I'm not sure if I did great in documenting that relativeAb returns counts per million, I'm open to learning more and making changes

File: .RData
Comment (lwaldron) [inline .RData:1]: It looks like .RData and .Rhistory are still here...

File: .gitignore
Comment (lwaldron) [inline .gitignore:1]: Adding .RData and .Rhistory to your .gitignore is a good idea for making sure you don't accidentally commit them in the future, but you'll still have to manually remove them (`git rm .Rhistory .RData`).

File: R/lefser.R
Comment (lwaldron) [inline R/lefser.R:1]: Looks good...

File: R/utils.R [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
+#'
+#'
 #' @description
 #' The function calculates the column totals and divides each value within the
 #' column by the respective column total.
+#' 
+#' 
 #'
-#' @param se A SummarizedExperiment object with counts
+#' @inheritParams lefser
```
Comment (lwaldron) [inline R/utils.R:11]: Can you explain the purpose of this change? (I'm not sure what it does myself)

File: R/utils.R [hunk 2]
```diff
@@ -1,12 +1,16 @@
 #' Utility function to calculate relative abundances
 #'
+#'
```
Comment (lwaldron) [inline R/utils.R:3]: Remove non-functional changes such as lines containing only `#'`, to simplify review.  In general if there are formatting or stylistic changes to make, it would be better to do those as their own separate PR.

File: R/utils.R [hunk 3]
```diff
@@ -19,60 +23,29 @@
...
-  
-  
-  res <- assay_data / csums * 1e6
-  
-  
-  cat("")
-  print(res)
-  
-  
+  div <- matrix(rep(csums, each = nrow(assay_data)), ncol = ncol(assay_data))
```
Comment (lwaldron) [inline R/utils.R:43]: Can you explain why you divide by `div` defined on line 43 instead of by `csums` as before? You might be right, I just don't understand the change!

File: R/utils.R [hunk 4]
```diff
@@ -19,60 +23,29 @@
...
 #' assay(relativeAb(se))
+#' 
+#' 
+#' utility function for relativeAb to return counts per million
+#'
+#' @description
+#' This function calculates the relative abundance of each feature in a dataset,
+#' expressed as counts per million (CPM) 
+#' 
+#' @param normalize Logical value indicating whether counts should be normalized
```
Comment (lwaldron) [inline R/utils.R]: You're adding documentation for an argument `normalize` but as far as I can see, `normalize` isn't currently an argument of the `relativeAb()` function. Am I missing something?

File: R/utils.R [hunk 5]
```diff
@@ -19,17 +23,29 @@
...
 #' assay(relativeAb(se))
+#' 
+#' 
+#' utility function for relativeAb to return counts per million
+#'
+#' @description
+#' This function calculates the relative abundance of each feature in a dataset,
+#' expressed as counts per million (CPM) 
+#' 
+#' @param Logical value indicating whether counts should be normalized
```
Comment (lwaldron) [inline R/utils.R:34]: What param (argument) is being documented on line 34?

**Review by lwaldron — CHANGES_REQUESTED** [review]
See some remaining comments within, getting closer!

File: .gitignore
Comment (Peacesandy) [reply .gitignore:1]: Okay, I'll work towards removing them in my next pull request

File: R/utils.R (hunk 2 above)
Comment (Peacesandy) [reply R/utils.R:3]: Okay, I'll make the changes

File: R/utils.R (hunk 4 above)
Comment (Peacesandy) [reply R/utils.R]: Okay, I'll make changes to the documentation

File: R/utils.R (hunk 1 above)
Comment (Peacesandy) [reply R/utils.R:11]: I think the @param is used in documenting the parameter [se] which is the summarized experiment. That part was already on the code, I didn't make any changes to it.

File: R/utils.R (hunk 2 above)
Comment (Peacesandy) [reply R/utils.R:3]: Done

File: R/utils.R (hunk 3 above)
Comment (Peacesandy) [reply R/utils.R:43]: I think the matrix div has the same dimension as assay_data

File: R/utils.R (hunk 5 above)
Comment (Peacesandy) [reply R/utils.R:34]: I just updated my the code by adding se as the param (argument)

I think the se parameter is the primary parameter controlling the behavior of this utility function. like, it returns a new SummarizedExperiment object with counts per million calculated and added as a new assay named rel_abs.

File: R/utils.R [hunk 6]
```diff
@@ -19,17 +19,26 @@
...
 #' assay(se)
 #' assay(relativeAb(se))
+#' 
+#' utility function for relativeAb to return counts per million
 #'
+#' @description
+#' This function calculates the relative abundance of each feature in the SummarizedExperiment 
+#' object containing count data, expressed as counts per million (CPM) 
+#' 
+#' @param se returns a new SummarizedExperiment object with counts per million
```
Comment (lwaldron) [inline R/utils.R:29]: Add a #' @param to document the assay argument: "Name or number of the assay of se containing count data to be used for the relative abundance calculation (default: 1L)."

Comment (lwaldron) [reply R/utils.R:29]: This is #' @returns (not param) to document what the function returns.

File: .RData
Comment (lwaldron) [inline .RData:1]: Remove .RData

**Review by lwaldron — CHANGES_REQUESTED** [review]
Remove .RData and make small changes to the relativeAb documentation.

File: R/utils.R [hunk 7]
```diff
@@ -19,17 +19,28 @@
...
 #' assay(se)
 #' assay(relativeAb(se))
+#' 
+#' utility function for relativeAb to return counts per million
 #'
+#' @description
+#' This function calculates the relative abundance of each feature in the SummarizedExperiment 
+#' object containing count data, expressed as counts per million (CPM) 
+#' 
+#' @returns se returns a new SummarizedExperiment object with counts per million
```
Comment (lwaldron) [inline R/utils.R:29]: Remove "se" from this text, since this describes the return value and not the parameter se

**Review by lwaldron — CHANGES_REQUESTED** [review]
Two more tiny comments, sorry about that!

File: R/utils.R [hunk 8]
```diff
@@ -19,17 +19,28 @@
 #' )
 #' assay(se)
 #' assay(relativeAb(se))
+#' 
+#' utility function for relativeAb to return counts per million
```
[the remark this answers is not in the document]
Comment (Peacesandy) [reply R/utils.R:23]: Okay

File: R/utils.R (hunk 7 above)
Comment (Peacesandy) [reply R/utils.R:29]: Okay, Thank you for the corrections

**Comment by Peacesandy** [comment]
Resolved all comments - I just pushed a commit

**Comment by lwaldron** [comment]
Great! Are test and check passing?

**Comment by Peacesandy** [comment]
> Great! Are test and check passing?

Yes they are

The output on the console

\```
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 0 ][ FAIL 0 | WARN 0 | SKIP 1 | PASS 0 ]

══ Skipped tests (1) ══════════════════════════════════════════════════════════════════
• identical(Sys.getlocale("LC_COLLATE"), "en_US.UTF-8") is not TRUE (1):
  test-lefser.R:6:3

[ FAIL 0 | WARN 0 | SKIP 1 | PASS 0 ]
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 4 ]> 
\```

**Review by lwaldron — APPROVED** [review]

**Comment by lwaldron** [comment]
I made 2 final changes:
1. deleted again the documentation of assay that I told you to add! It was actually covered by @inheritParams
2. fixed the unit tests that were only done on US locale, since results change now that we use CPM instead of fraction.  The numeric tests should ideally be changed to be done on any locale.

**Comment by lwaldron** [comment]
Great work @Peacesandy !!!

**Comment by Peacesandy** [comment]
Thank you so much @lwaldron I really learnt a lot
